[jit] register distributed backward

wanchaol opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #38657 some refactor on register_distributed_ops
* **#38494 [jit] register distributed backward**
* #38656 [dist_autograd] expose distributed backward C++ API

This register distributed.autograd.backward to jit

Differential Revision: [D21596133](https://our.internmc.facebook.com/intern/diff/D21596133)

wanchaol committed:
[jit] register distributed backward

This register distributed.autograd.backward to jit

[ghstack-poisoned]

wanchaol committed:
Update on "[jit] register distributed backward"


This register distributed.autograd.backward to jit

[ghstack-poisoned]

wanchaol committed:
Update on "[jit] register distributed backward"


This register distributed.autograd.backward to jit

[ghstack-poisoned]

wanchaol committed:
Update on "[jit] register distributed backward"


This register distributed.autograd.backward to jit

[ghstack-poisoned]

wanchaol committed:
Update on "[jit] register distributed backward"


This register distributed.autograd.backward to jit

[ghstack-poisoned]

xush6528 commented on torch/csrc/jit/runtime/register_distributed_ops.cpp:
```diff
@@ -39,6 +42,10 @@ c10::AliasAnalysisKind aliasAnalysisFromSchema() {
+c10::AliasAnalysisKind aliasAnalysisConservative() {
+  return c10::AliasAnalysisKind::CONSERVATIVE;
+}
+
```
Re-use the one defined in torch/csrc/jit/runtime/register_ops_utils.h?

wanchaol replied:
did this in https://github.com/pytorch/pytorch/pull/38657

xush6528 commented on torch/csrc/jit/runtime/register_distributed_ops.cpp:
```diff
@@ -110,6 +117,28 @@ RegisterOperators reg_rpc_ops(
+           }
+           return 0;
+         },
+         aliasAnalysisConservative()),
```
Will this be a concern that we have duplicate code for pybind and builtin op? 
Can defining it as a custom op helps since it can make single copy of C++ code be called from both python and torchscript?

wanchaol replied:
exposed the C++ API instead in https://github.com/pytorch/pytorch/pull/38656, so we share impls between both

wanchaol committed:
Update on "[jit] register distributed backward"


This register distributed.autograd.backward to jit

Differential Revision: [D21596133](https://our.internmc.facebook.com/intern/diff/D21596133)

[ghstack-poisoned]

wanchaol committed:
Update on "[jit] register distributed backward"


This register distributed.autograd.backward to jit

Differential Revision: [D21596133](https://our.internmc.facebook.com/intern/diff/D21596133)

[ghstack-poisoned]

wanchaol committed:
Update on "[jit] register distributed backward"


This register distributed.autograd.backward to jit

Differential Revision: [D21596133](https://our.internmc.facebook.com/intern/diff/D21596133)

[ghstack-poisoned]

wanchaol committed:
Update on "[jit] register distributed backward"


This register distributed.autograd.backward to jit

Differential Revision: [D21596133](https://our.internmc.facebook.com/intern/diff/D21596133)

[ghstack-poisoned]

wanchaol committed:
Update on "[jit] register distributed backward"


This register distributed.autograd.backward to jit

Differential Revision: [D21596133](https://our.internmc.facebook.com/intern/diff/D21596133)

[ghstack-poisoned]

wanchaol committed:
Update on "[jit] register distributed backward"


This register distributed.autograd.backward to jit

Differential Revision: [D21596133](https://our.internmc.facebook.com/intern/diff/D21596133)

[ghstack-poisoned]

wanchaol committed:
Update on "[jit] register distributed backward"


This register distributed.autograd.backward to jit

Differential Revision: [D21596133](https://our.internmc.facebook.com/intern/diff/D21596133)

[ghstack-poisoned]

wanchaol committed:
Update on "[jit] register distributed backward"


This register distributed.autograd.backward to jit

Differential Revision: [D21596133](https://our.internmc.facebook.com/intern/diff/D21596133)

[ghstack-poisoned]

wanchaol committed:
Update on "[jit] register distributed backward"


This register distributed.autograd.backward to jit

Differential Revision: [D21596133](https://our.internmc.facebook.com/intern/diff/D21596133)

[ghstack-poisoned]

wanchaol committed:
Update on "[jit] register distributed backward"


This register distributed.autograd.backward to jit

Differential Revision: [D21596133](https://our.internmc.facebook.com/intern/diff/D21596133)

[ghstack-poisoned]

wanchaol committed:
Update on "[jit] register distributed backward"


This register distributed.autograd.backward to jit

Differential Revision: [D21596133](https://our.internmc.facebook.com/intern/diff/D21596133)

[ghstack-poisoned]
